### ISSUE TITLE
SfePy data generation errors

### ISSUE THREAD

**Issue by amarshall40** [materialsinnovation/pymks#397]
When I try to use SfePy data generation functions (make_elastic_FE_strain_random, make_elastic_stiffness, etc.), an error is returned. I'm using a Conda installation. Example from intro.ipynb:

from pymks.datasets import make_elastic_stress_random

grain_size = [(47, 6), (4, 49), (14, 14)]
n_samples = [200, 200, 200]

X_train, y_train = make_elastic_stress_random(n_samples=n_samples, size=(51, 51),
                                                  grain_size=grain_size, seed=0)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-87143f67ce67> in <module>()
      5 
      6 X_train, y_train = make_elastic_stress_random(n_samples=n_samples, size=(51, 51),
----> 7                                                   grain_size=grain_size, seed=0)

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pymks\datasets\__init__.py in make_elastic_stress_random(n_samples, elastic_modulus, poissons_ratio, size, macro_strain, grain_size, seed, volume_fraction, percent_variance)
    302     X_cal, y_cal = make_elastic_FE_strain_delta(elastic_modulus,
    303                                                 poissons_ratio, size,
--> 304                                                 macro_strain)
    305     n_states = len(elastic_modulus)
    306     basis = DiscreteIndicatorBasis(n_states)

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pymks\datasets\__init__.py in make_elastic_FE_strain_delta(elastic_modulus, poissons_ratio, size, macro_strain)
     42 
     43     X = make_delta_microstructures(len(elastic_modulus), size=size)
---> 44     FEsim.run(X)
     45     return X, FEsim.response
     46 

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pymks\datasets\elastic_FE_simulation.py in run(self, X)
    195         stress = []
    196         for x in X_property:
--> 197             strain_, displacement_, stress_ = self._solve(x)
    198             strain.append(strain_)
    199             displacement.append(displacement_)

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pymks\datasets\elastic_FE_simulation.py in _solve(self, property_array)
    520         ls = ScipyDirect({})
    521 
--> 522         pb = Problem('elasticity', equations=eqs, auto_solvers=None)
    523 
    524         pb.time_update(

TypeError: __init__() got an unexpected keyword argument 'auto_solvers'

**Comment by amarshall40** [comment]
Update: I'm making modifications to my local version of PyMKS. I took out the auto_solvers keyword and I no longer get the error listed in the previous post. However, I'm getting a new error chain:

Traceback (most recent call last):

  File "<ipython-input-31-183d9a8e9b04>", line 1, in <module>
    runfile('C:/Users/amarshall40/Documents/PythonScripts/pymks_intro.py', wdir='C:/Users/amarshall40/Documents/PythonScripts')

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\spyder\utils\site\sitecustomize.py", line 705, in runfile
    execfile(filename, namespace)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\spyder\utils\site\sitecustomize.py", line 102, in execfile
    exec(compile(f.read(), filename, 'exec'), namespace)

  File "C:/Users/amarshall40/Documents/PythonScripts/pymks_intro.py", line 43, in <module>
    grain_size=grain_size, seed=0)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\pymks\datasets\__init__.py", line 304, in make_elastic_stress_random
    macro_strain)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\pymks\datasets\__init__.py", line 44, in make_elastic_FE_strain_delta
    FEsim.run(X)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\pymks\datasets\elastic_FE_simulation.py", line 197, in run
    strain_, displacement_, stress_ = self._solve(x)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\pymks\datasets\elastic_FE_simulation.py", line 538, in _solve
    vec = pb.solve()

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\sfepy\discrete\problem.py", line 1372, in solve
    self.time_update(tss.ts) # Only having adi is required here(?)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\sfepy\discrete\problem.py", line 664, in time_update
    functions, create_matrix, is_matrix)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\sfepy\discrete\problem.py", line 617, in update_equations
    active_only=ac)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\sfepy\discrete\equations.py", line 306, in time_update
    active_only=active_only)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\sfepy\discrete\variables.py", line 430, in equation_mapping
    problem=problem)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\sfepy\discrete\variables.py", line 1379, in equation_mapping
    problem=problem, warn=warn)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\sfepy\discrete\common\dof_info.py", line 412, in map_equations
    fun = get_condition_value(bc.match, functions, 'EPBC', bc.name)

  File "C:\Users\amarshall40\AppData\Local\Continuum\anaconda3\lib\site-packages\sfepy\discrete\conditions.py", line 27, in get_condition_value
    raise ValueError('no functions given for %s %s!' % (kind, name))

ValueError: no functions given for EPBC periodic_x!

It appears that the functions variable of the class sfepy.discrete.functions.Functions is being set to None somewhere in the workflow, which eventually causes the ValueError. I'm having trouble locating the source of the None type assignment, so any assistance would be greatly appreciated.

**Comment by wd15** [comment]
I think I would recommend that you use the develop branch at this point alongside version 2018.2 version of Sfepy. The Conda recipe is now probably out of date with master unfortunately.

**Comment by wd15** [comment]
What version of Sfepy are you currently using? Master probably hasn't been updated to use later versions.

**Comment by amarshall40** [comment]
I have SfePy 2018.2 so yes, I'm guessing that the PyMKS I installed via Conda (0.3.4) is the source of the problem


Andrew Marshall

Graduate Research Assistant
amarshall40@gatech.edu


________________________________

**Comment by wd15** [comment]
The problem is that we don't fix versions of dependencies in the conda-forge recipe as you're not supposed to do that for conda-forge. 

Also try the Nix install if you like with develop. I'm always on top of that as I use it on a daily basis.

**Comment by amarshall40** [comment]
I'm not super experienced with GitHub, so I hope you'll allow me to ask a dumb question - how do I install from the develop branch or use the Nix install? I tried installing using the git URL:

https://github.com/materialsinnovation/pymks.git

I installed PyMKS (0.3.4.dev5+g4a78f56) using this but I get the same errors as before - is this the Master version? The git URL appears the same whether I've selected the Master branch or the develop branch.

Thanks again!

**Comment by wd15** [comment]
## Using PyMKS with an existing Python env

To use the develop branch in a Python environment that you have already.

    $ conda activate my-python-env
    $ git clone https://github.com/materialsinnovation/pymks.git
    $ cd pymks
    $ git checkout develop
    $ python setup.py develop

You should now be able to develop in your working pymks copy and have the changes reflected when you do `import pymks`.

To check which branch you're actually using use,

    $ git branch
    * develop
      master

This [small course](http://swcarpentry.github.io/git-novice/) might help if you are just getting started with Git.
    
## Using PyMKS with Nix

### Get Nix Working

To use the Nix installation (works only on Mac and Linux) use [these notes](https://github.com/wd15/nixes/blob/master/NIX-NOTES.md). Essentially, you should only need to do,

    $ bash <(curl https://nixos.org/nix/install)

to get going with Nix. You should be able to do 

    $ nix-env -i hello
    $ hello
    Hello, world!

to check that things are working.

### Get PyMKS working with Nix

Once you have a working Nix installation go to the base PyMKS directory (where `default.nix` is) and run

    $ nix-shell --pure

That's it. You are now in the Nix shell. You should be able to run the PyMKS tests at this point

    $ py.test

and all the test should pass. Maybe 1 or 2 failures due to issues we are trying to fix..

**Comment by wd15** [comment]
@amarshall40, the following should get you a working version of the latest release of pymks in conda. At least for me, it passed all the tests.

    $ conda create -n pymks
    $ conda activate pymks
    $ conda install pymks scikit-learn=0.18 sfepy=2017.2 nomkl
 
To test that everything works, use

    $ python -c "import pymks; pymks.test()"

This is a non-developer installation. To make this a develop installation, Git clone the repo and run `python setup.py devolep`.

**Comment by amarshall40** [comment]
I followed the instructions for the develop branch in your previous post, and it looks like the installation was successful:

![image](https://user-images.githubusercontent.com/51411381/59309688-3da69b00-8c72-11e9-8bff-b3519e3c4707.png)

![image](https://user-images.githubusercontent.com/51411381/59309667-2ebfe880-8c72-11e9-998b-6ac64a8f5bae.png)

However, when I import PyMKS and attempt to access SfePy, I get the error from the first post in this thread. Sample code:

![image](https://user-images.githubusercontent.com/51411381/59309872-bf96c400-8c72-11e9-8e69-ece9363af9b4.png)

Error:
![image](https://user-images.githubusercontent.com/51411381/59309899-cf160d00-8c72-11e9-9654-eb012d894f6c.png)

Am I missing a step? I'm not sure at the moment why I'd be getting this error if I'm using the develop branch version of PyMKS and everything is up to date.

**Comment by wd15** [comment]
Which versions of Scikit-learn and Sfepy are you using? The PyMKS develop branch is currently tested against Sfepy version 2018.2 and Scikit-learn version 0.19.2. If you are using Conda, you should be able to do 

     $ conda install scikit-learn=0.19 sfepy=2018.2 nomkl

to get things working.

**Comment by amarshall40** [comment]
I updated scikit-learn from 0.19.1 to 0.19.2, so now my modules look like this:

![image](https://user-images.githubusercontent.com/51411381/59363541-5612c600-8d03-11e9-83b8-e2e416f03669.png)
![image](https://user-images.githubusercontent.com/51411381/59363557-59a64d00-8d03-11e9-897b-20896d2dd0be.png)
![image](https://user-images.githubusercontent.com/51411381/59363563-5ca13d80-8d03-11e9-8d0c-dfcbfad42d5e.png)

However, I'm still getting the same error as before.

Example code:

![image](https://user-images.githubusercontent.com/51411381/59363619-72166780-8d03-11e9-949b-d525c8c8e595.png)

Error:

![image](https://user-images.githubusercontent.com/51411381/59363628-7773b200-8d03-11e9-8fe6-2d6df07c88f6.png)

**Comment by wd15** [comment]
Here goes, let's start with a fresh Conda environment and go from there. Start over and follow these commands

    $ conda create -n pymks-test
    $ conda activate pymks-test
    $ conda install pymks scikit-learn=0.19 sfepy=2018.2 nomkl toolz dask scipy=0.19.1
   
Now use the pymks develop branch

    $ cd .../pymks
    $ git checkout develop
    $ python setup.py develop

Now run the tests 

    $ python -c "import pymks; pymks.test()"

All the tests should pass (they do for me). Please report if they do not. Check the versions of the libraries are correct.

    $ python -c "import scipy; print(scipy.__version__)"
    $ python -c "import sklearn; print(sklearn.__version__)"
    $ python -c "import pymks; print(pymks.__version__)"
    $ python -c "import sfepy; print(sfepy.__version__)"

Please report the above to ensure that the Python run time environment has the correct versions. Now your example above should work.

**Comment by amarshall40** [comment]
I'm actually having some strange issues with Conda on my desktop now - I don't think it's anything you suggested, but I'm reinstalling so I can start over fresh. Once I get things back to normal I'll try your latest suggestion and let you know how it goes.

**Comment by amarshall40** [comment]
Ok, I got my Conda situation back to normal and set up the environment as you specified.

Here's what I get in terms of versions:

![image](https://user-images.githubusercontent.com/51411381/59470909-48466900-8e07-11e9-8c83-d7a6a73e4348.png)

For PyMKS the develop branch appears to be selected when looked up via conda list:

![image](https://user-images.githubusercontent.com/51411381/59471036-cacf2880-8e07-11e9-8517-5648e2270706.png) 

I tested out the new environment in Spyder and on Jupyter notebook, ensuring that the correct environment was being used. Spyder:

![image](https://user-images.githubusercontent.com/51411381/59470980-83489c80-8e07-11e9-8882-d3b72a39bbd8.png)

However, I'm still getting the same error as before unfortunately:

![image](https://user-images.githubusercontent.com/51411381/59471001-9d827a80-8e07-11e9-9014-3dd07d1d79fa.png)

Any other thoughts?

**Comment by wd15** [comment]
What happens when you run the tests?

**Comment by amarshall40** [comment]
Ah, sorry. Yes, I'm getting errors:

![image](https://user-images.githubusercontent.com/51411381/59471510-47163b80-8e09-11e9-8606-b1ec3d05ec27.png)

Any idea how to address this?

**Comment by wd15** [comment]
@amarshall40, I have been away for the last week. Let's try again. Looking over your above [comment](https://github.com/materialsinnovation/pymks/issues/397#issuecomment-501899592), it seems that you have version 1.2.1 of Scipy, but in the Conda command I gave you, it fixes Scipy to be 0.19.1. That doesn't make sense to me. Can you try running the Conda commands in the [comment](https://github.com/materialsinnovation/pymks/issues/397#issuecomment-501425048) exactly as they are?

Also, the [comment](https://github.com/materialsinnovation/pymks/issues/397#issuecomment-501902086) with your test results doesn't show the failures. Can you pipe the results to a file an upload the file rather than using an image?

You might also try doing this outside of the Spyder environment which could be confusing things.

**Comment by wd15** [comment]
Also, looking at the [comment where you have the error](https://github.com/materialsinnovation/pymks/issues/397#issuecomment-501899592), the code that is listed is from [master](https://github.com/materialsinnovation/pymks/blob/master/pymks/datasets/elastic_FE_simulation.py#L522). The develop version of pymks no longer has that code (`auto_solvers=None`), see [here](https://github.com/materialsinnovation/pymks/blob/develop/pymks/datasets/elastic_FE_simulation.py#L522). So, it seems to me that the versions of the codes listed at your command line are not the same versions of the codes you are using in Spyder.

**Comment by beyucel** [comment]
Hi, Andrew. I tried the following installation for Windows and all the notebooks are working with the following installation for **master** branch: 

Start with a fresh Conda environment and go from there. Then install the pymks with specific versions of sfepy, matplotlib,scipy etc. 

    $ conda create -n pymks-test1
    $ conda activate pymks-test1
    $ conda install -c conda-forge pymks scikit-learn=0.19 sfepy=2017.2 toolz scipy=0.19.1 matplotlib=2.0.2

After this installation, jupyter notebooks should work for the master branch. For that installation, I used **Anaconda Prompt** on windows. I would suggest you to use Anaconda Prompt as well
